### PR TITLE
Documentation and code clean up

### PR DESCRIPTION
**Pull request by marwanrahman**
- Updated readme
- Rearranging old code for optical flow data processing methods

### CODE REVIEW

**Review by bvanber — APPROVED** [review]

File: src/preprocessor_flow.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+    x = tf.map_fn(
+        lambda x1: tf.image.random_brightness(x1, max_delta=cfg['TRAIN']['PARAMS']['AUGMENTATION']['BRIGHTNESS_DELTA']),
+        x)
+    x = tf.map_fn(lambda x1: tf.image.random_contrast(x1, cfg['TRAIN']['PARAMS']['AUGMENTATION']['CONTRAST_BOUNDS'][0],
+                                                      cfg['TRAIN']['PARAMS']['AUGMENTATION']['CONTRAST_BOUNDS'][1]), x)
+    x = tf.map_fn(lambda x1: preprocessor.random_shift_clip(x1), x)
+    x = tf.map_fn(lambda x1: preprocessor.random_flip_left_right_clip(x1), x)
+    x = tf.map_fn(lambda x1: preprocessor.random_zoom_clip(x1), x)
+    # AUGMENT RANDOM NOISE WONT WORK RN BECAUSE
```
Comment (bvanberl) [inline src/preprocessor_flow.py:27]: Did we want to remove this comment for now?

File: src/preprocessor_flow.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    #x = tf.map_fn(lambda x1: random_shift_clip(x1), x)
+    x = tf.map_fn(lambda x1: preprocessor.random_flip_left_right_clip(x1), x)
+    #x = tf.map_fn(lambda x1: random_zoom_clip(x1), x)
+    x = tf.map_fn(lambda x1: preprocessor.random_noise(x1), x)
+
+    # Separate regular and flow mini-clips
+    x1, x2 = tf.split(x, 2, axis=-4)
+    x2 = tf.split(x2, [2, 1], axis=-1)[0]
+
+    #tf.ensure_shape(x1, [window] + img_size + [3])
```
Comment (bvanberl) [inline src/preprocessor_flow.py:63]: I assume we can remove these commented lines.

File: README.md [hunk 3]
```diff
@@ -27,28 +27,49 @@ of predicting the presence (positive class) or absence (negative class) of lung
    (.env) > pip install -r requirements.txt
    ```
```
Comment (bvanberl) [inline README.md:28]: For some reason I can't comment on line 12. But "Extraneous Data" shows up on the same line as "Building a Dataset". Try adding spaces at the end of line 12.

File: README.md [hunk 4]
```diff
@@ -27,28 +27,49 @@ of predicting the presence (positive class) or absence (negative class) of lung
...
 
 For all clips contained in our internal database, a single call to [_run.py_](src/data/run.py)
-will generate mini-clips in NPZ (serialized NumPy array) form, as required for the training pipeline. Before running this
-file, ensure that the parameters under _PREPROCESS >> PARAMS_ in [_config.yml_](config.yml) are 
-set as desired. Additionally, CSVs containing pleural region ROIs are needed, and can be 
-downloaded from the appropriate source (Deep Breathe -> lung_sliding_classifier -> data -> 
-new_csvs -> Boxes). [run.py](src/data/run/py) will first call [_download_videos.py_](src/data/download_videos.py) to download appropriate videos from the database, then [_mask.py_](src/data/mask.py) is 
+will generate M-mode images in NPZ format, as required for the training pipeline. Before running this
+file, ensure that the parameters under _`PREPROCESS` >> `PARAMS`_ in [_config.yml_](config.yml) are 
+set as desired. Additionally, the following CSVs must be placed in the _`PREPROCESS` >> `PATHS` >> `CSVS_OUTPUT`_:
```
Comment (bvanberl) [inline README.md:57]: I wonder if we can rearrange the information presented in this section as a list of steps. E.g. step 1 would be setting the proper config values. Step 2 is getting those CSVs in the right place, etc.

File: README.md [hunk 5]
```diff
@@ -27,28 +27,49 @@ of predicting the presence (positive class) or absence (negative class) of lung
...
 [_smoothing_filter.py_](src/data/smoothing_filter.py), which applies a median filter to masked
 clips. Finally, [_to_npz.py_](src/data/to_npz.py) will be called to convert masked (and optionally smoothed) 
-clips to mini-clips with fixed length, saved in the NPZ format. 
+clips to sub-clips with fixed length, construct M-mode images for each sub-clip, and save the M-mode images in the NPZ 
+format. 
+
+Note: Some additional preprocessing methods exist from early experimentation for this project. This includes using 
+optical flow instead of M-Mode images, which can be done by setting `PREPROCESS` >> `PARAMS` >> `FLOW`.
 
 ### Extraneous data
```
Comment (bvanberl) [inline README.md:74]: Can we replace this with "Additional Data" or something like that? The word extraneous implies irrelevance.

File: README.md [hunk 6]
```diff
@@ -68,13 +89,175 @@ masks extraneous videos. Finally, [_extra_clips_to_npz.py_](src/data/extras/extr
 should be called to produce mini-clips (with associated metadata) as NPZs. Note that smoothing 
 (median filtering) as a preprocessing step is not currently implemented for extraneous data. 
 
+## Use Cases
+
+### Model Training
+To train a model that is already defined in this repository:
```
Comment (bvanberl) [inline README.md:95]: "...model whose architecture is already defined..."

File: README.md [hunk 7]
```diff
@@ -68,13 +89,175 @@ masks extraneous videos. Finally, [_extra_clips_to_npz.py_](src/data/extras/extr
...
+wish to train, and `TRAIN` >> `EXPERIMENT_TYPE` to _hparam_search_. 
+4. Set the associated model hyperparameter values based on the chosen model definition (in `TRAIN` >> `PARAMS`).
+5. Set the hyperparameter types and ranges to vary for optimization in `HPARAM_SEARCH` >> `MODEL_DEF`, where _MODEL_DEF_ 
+is the same model chosen in `TRAIN` >> `MODEL_DEF`.
+6. Set the number of runs to complete for the optimization in `HPARAM_SEARCH` >> `N_EVALS`, as well as other configurable
+parameters.
+7. Run [train.py](src/train.py).
+8. View logs and saved models in the directory specified in `TRAIN` >> `PATHS` >> `TENSORBOARD` and `MODEL_OUT`. Logs 
+can be viewed in TensorBoard by running `tensorboard --logdir path` in terminal, where path is the directory containing 
+the logs.
```
Comment (bvanberl) [inline README.md:151]: Can we add a note on where the spreadsheet and pictorial results are stored?

File: README.md [hunk 8]
```diff
@@ -68,13 +89,175 @@ masks extraneous videos. Finally, [_extra_clips_to_npz.py_](src/data/extras/extr
...
+
+### Bayesian Hyperparameter Optimization with Integrated Cross-Validation
+To increase confidence in the model’s generalizability for each hyperparameter combination, each run in this search 
+consists of training the model a number of times on the same hyperparameter combination but on different training and 
+validation data splits. The average validation set performance across training runs on the same hyperparameter combination
+is used as the result for each hyperparameter search run.
```
Comment (bvanberl) [inline README.md:157]: Do you think that we can combine this section and the one above it? Perhaps a paragraph introducing the optimization and then the 2 sets of steps. Or adding conditional instructions to each step depending on whether you are running with cross val.

File: README.md [hunk 9]
```diff
@@ -68,13 +89,175 @@ masks extraneous videos. Finally, [_extra_clips_to_npz.py_](src/data/extras/extr
...
+For user convenience, the config file is organized into major components
+of the model development pipeline. Many fields do not need to be modified by
+the typical user, but others may be modified to suit the user's specific
+goals.
 
-To train a custom model, please follow the style laid out in [models.py](src/models/models.py) and place the code in that file. Furthermore, please add to the config file by following the style in _TRAIN >> PARAMS_ to add configurable options, such as the learning rate. Although you don't _have_ to make use of these parameters, please at least create this section in [config.yml](config.yml) with the same name in capitals as you decided in [models.py](src/models/models.py).
+## Project Structure
+The project looks similar to the directory structure below. Disregard
+any _.gitkeep_ files, as their only purpose is to force Git to track
+empty directories. Disregard any _.\_init\_.py_ files, as they are
```
Comment (bvanberl) [inline README.md:212]: I don't think we have any \_\_init\_\_.py files in this particular project

**Review by bvanberl — APPROVED** [review]
